### ISSUE TITLE
Allow specifying generics at the function call invocation.

### ISSUE THREAD

**Issue by Raynos** [giltayar/types-as-comments-proposal#3]
When reading the generics section of the README is noticed that there was no example of

```js
const m = new Map<string, string>()
```

One of the common limitations I run into with JSDoc and the typescript language is not being able to specify the generics.

However, since you've added a cast syntax this can be worked around

```js
const m = new Map() as Map<string, string>
```

**Comment by Raynos** [comment]
Hmm, after thinking about it a little more, I believe that `new Map<string, string>` is actually a different cast syntax in typescript anyway, it's the same as `new Map() as Map<string, string>`.

Maybe a section in "Up for debate" about adding this "syntax sugar" is sufficient.

**Comment by giltayar** [comment]
@Raynos Good point!  I have the same problem with JSDoc (specifically in Fastify) and solve it using `@type`, which is essentially casting. 

But asking devs to use casting at the call site is, I believe, to much to ask. On the other hand, enabling generics at the expression level is stepping into uncharted territory, especially given that `<` is a valid operator in an expression.

BTW, I believe your code is more correct this way:

```ts
new (Map as Map<string, string>)()
```

No solutions yet, just thoughts. @Raynos: do you think we should leave it as is and _not_ enable call site generics?

**Comment by Raynos** [comment]
I think adding a section to the README about "call site generics" not being in the proposal and recommending casting instead is sufficient for now.

**Comment by giltayar** [comment]
I checked a bit more, and using casting is _not_ an option: 

```ts
function foo<T>(t: T): T {return t}

// This is a syntax error
(foo as foo<number>)(4)
```

I continued checking, and found out that:

```ts
// This is an error
const z = foo<number>;

// and this is not an error
(foo)<number>(4)

// Also this is not an error
const foos = [foo];
foos[0]<number>(4);
```

Not exactly sure how TS parses it, but it would be hell to write down in the ES standard, I think. I'm not a parsing expert.


Worse, look at this!

```ts
const bar = <T>(t: T) => t

bar(5)
```

Now we have the `<T>` without anything before it!

So, yes, I agree with you about adding a section to the proposal about this, and wait for a fuller understanding of how a parser should deal with this.

@littledan, any thoughts on this? I totally missed this when going over the TS constructs.
